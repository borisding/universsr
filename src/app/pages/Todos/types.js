export const FETCH_TODO_BEGIN = 'FETCH_TODO_BEGIN';
export const FETCH_TODO_SUCCESS = 'FETCH_TODO_SUCCESS';
export const FETCH_TODO_FAILURE = 'FETCH_TODO_FAILURE';
export const ADD_TODO = 'ADD_TODO';
export const UPDATE_TODO = 'UPDATE_TODO';
