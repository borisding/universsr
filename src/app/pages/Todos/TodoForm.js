import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { TextInput } from '@common/components';
import './styles/TodoForm.scss';

const initialState = {
  todoInput: '',
  isInvalid: false
};

export default class TodoForm extends Component {
  static propTypes = {
    addTodo: PropTypes.func.isRequired
  };

  constructor() {
    super();
    this.state = initialState;
  }

  handleChange = evt => {
    this.setState({
      todoInput: evt.target.value,
      isInvalid: initialState.isInvalid
    });
  };

  handleBlur = evt => {
    if (this.state.isInvalid) {
      this.setState({
        todoInput: evt.target.value,
        isInvalid: !this.state.isInvalid
      });
    }
  };

  handleSubmit = evt => {
    evt.preventDefault();
    const { todoInput } = this.state;

    if (todoInput.trim() === '') {
      return this.setState({
        todoInput,
        isInvalid: !initialState.isInvalid
      });
    }

    return this.props
      .addTodo(todoInput)
      .then(() => this.setState(initialState));
  };

  render() {
    const { todoInput, isInvalid } = this.state;
    const styles = classNames('todo-input', { 'error-input': !!isInvalid });
    const placeholder = isInvalid ? 'Todo is required!' : 'Enter new todo...';

    return (
      <form onSubmit={this.handleSubmit}>
        <TextInput
          name="newTodo"
          value={todoInput}
          styleName={styles}
          placeholder={placeholder}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />
      </form>
    );
  }
}
