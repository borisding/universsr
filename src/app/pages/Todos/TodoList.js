import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from '@common/components';
import './styles/TodoList.scss';

export default class TodoList extends Component {
  static propTypes = {
    todos: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        todo: PropTypes.string.isRequired,
        done: PropTypes.bool.isRequired
      })
    ).isRequired,
    updateTodo: PropTypes.func.isRequired
  };

  isTruthy(value) {
    return value === true;
  }

  render() {
    const { todos, updateTodo } = this.props;
    const totalDone = todos.filter(todo => this.isTruthy(todo.done)).length;

    return (
      <div styleName="todos-container">
        <h3>{`My Todos (${totalDone} / ${todos.length})`}</h3>
        <ul>
          {todos.map(record => (
            <li key={record.id}>
              <label htmlFor={record.id}>
                <Checkbox
                  id={record.id}
                  value={record.id}
                  onChange={evt => updateTodo(evt.target)}
                  isChecked={this.isTruthy(record.done)}
                />
                <span styleName={!record.done ? 'pending' : 'done'}>
                  {record.todo}
                </span>
              </label>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}
