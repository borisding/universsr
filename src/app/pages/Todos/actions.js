import uuidv4 from 'uuid/v4';
import { service } from '@utils';
import * as types from './types';

// async/await example of making todo api request
export const fetchTodos = () => {
  return async (dispatch, getState, { requestError }) => {
    try {
      await dispatch({ type: types.FETCH_TODO_BEGIN });
      const res = await service.get('/todos');

      return dispatch({
        type: types.FETCH_TODO_SUCCESS,
        payload: res.data
      });
    } catch (err) {
      dispatch({ type: types.FETCH_TODO_FAILURE });
      dispatch(requestError(err));
    }
  };
};

// pre-fetch todos when there is no populated todos in store
export const prefetchTodos = () => (dispatch, getState) => {
  const { isFetching, isFetched } = getState().todos;

  // no action when it's fetching todos
  // or, todos is already fetched
  if (isFetching || isFetched) {
    return Promise.resolve(true);
  }

  // else, just proceed to fetching new todos list
  return dispatch(fetchTodos());
};

// fake adding new todo without saving into db
export const addTodo = input => {
  return dispatch =>
    Promise.resolve(
      dispatch({
        type: types.ADD_TODO,
        payload: { id: uuidv4(), todo: input, done: false }
      })
    );
};

// fake updating todo status without saving into db
export const updateTodo = ({ value, checked }) => ({
  type: types.UPDATE_TODO,
  payload: { id: value, done: checked }
});
