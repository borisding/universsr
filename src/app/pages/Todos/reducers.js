import * as types from './types';

const initialState = {
  isFetching: false,
  isFetched: false,
  data: []
};

const updateTodo = (state, action) =>
  state.data.map(record => {
    if (record.id !== action.payload.id) {
      return record;
    }

    return { ...record, done: action.payload.done };
  });

export default (state = initialState, action) => {
  switch (action.type) {
    case types.FETCH_TODO_BEGIN:
      return Object.assign({}, state, { isFetching: true, isFetched: false });

    case types.FETCH_TODO_FAILURE:
      return Object.assign({}, state, { isFetching: false, isFetched: false });

    case types.FETCH_TODO_SUCCESS:
      return {
        isFetching: false,
        isFetched: true,
        data: [...state.data, ...action.payload]
      };

    case types.ADD_TODO:
      return { ...state, data: [...state.data, action.payload] };

    case types.UPDATE_TODO:
      return { ...state, data: [...updateTodo(state, action)] };

    default:
      return state;
  }
};
