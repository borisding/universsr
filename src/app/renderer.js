import React from 'react';
import DocumentTitle from 'react-document-title';
import serialize from 'serialize-javascript';
import flushChunks from 'webpack-flush-chunks';
import { renderToString } from 'react-dom/server';
import { flushChunkNames } from 'react-universal-component/server';
import { matchRoutes, renderRoutes } from 'react-router-config';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import storeFactory from './store';
import routes from './routes';

// preload data for matched route
function prefetchBranchData(store, url) {
  try {
    const branch = matchRoutes(routes, url);
    const promises = branch.map(({ route, match }) => {
      const { loadData } = route;
      const { dispatch } = store;

      if (match && match.isExact && loadData) {
        return Array.isArray(loadData)
          ? Promise.all(loadData.map(action => dispatch(action(match))))
          : dispatch(loadData(match));
      }

      return Promise.resolve(null);
    });

    return Promise.all(promises);
  } catch (err) {
    throw err;
  }
}

// export default server renderer and receiving stats
// also, should allow it to be mounted as middleware for production usage
export default function serverRenderer({ clientStats }) {
  return async (req, res, next) => {
    try {
      const context = {};
      const store = storeFactory();

      await prefetchBranchData(store, req.url);

      const appString = renderToString(
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            {renderRoutes(routes)}
          </StaticRouter>
        </Provider>
      );

      const pageTitle = DocumentTitle.rewind();
      const chunksOptions = { chunkNames: flushChunkNames() };
      const { scripts, stylesheets } = flushChunks(clientStats, chunksOptions);

      const preloadedState = serialize(store.getState(), { isJSON: true });
      const { statusCode = 200, redirectUrl } = context;
      const { nonce } = res.locals;

      if ([301, 302].includes(statusCode) && redirectUrl) {
        return res.redirect(statusCode, redirectUrl);
      }

      res.status(statusCode);
      res.render('index', {
        pageTitle,
        appString,
        preloadedState,
        stylesheets,
        scripts,
        nonce
      });
    } catch (err) {
      next(new Error(err));
    }
  };
}
