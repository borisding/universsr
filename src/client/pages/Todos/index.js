import lazy from '@client/pages/lazy';

export default lazy(() => import(/* webpackChunkName: 'todos' */ './Todos'));
