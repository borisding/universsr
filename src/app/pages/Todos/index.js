import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { PageTitle, Loader } from '@common/components';
import TodoForm from './TodoForm';
import TodoList from './TodoList';
import * as todoActions from './actions';

class Todos extends Component {
  static propTypes = {
    todos: PropTypes.arrayOf(PropTypes.object).isRequired,
    actions: PropTypes.objectOf(PropTypes.func).isRequired
  };

  componentDidMount() {
    this.props.actions.prefetchTodos();
  }

  renderTodos = () => {
    const { isFetching, actions, todos } = this.props;

    if (isFetching) {
      return <Loader />;
    }

    return (
      <Fragment>
        <TodoForm addTodo={actions.addTodo} />
        <hr />
        <TodoList todos={todos} updateTodo={actions.updateTodo} />
      </Fragment>
    );
  };

  render() {
    return (
      <PageTitle title="Todos">
        <h3>Todos Demo</h3>
        {this.renderTodos()}
      </PageTitle>
    );
  }
}

export default connect(
  state => ({ todos: state.todos.data, isFetching: state.todos.isFetching }),
  dispatch => ({ actions: bindActionCreators(todoActions, dispatch) })
)(Todos);
