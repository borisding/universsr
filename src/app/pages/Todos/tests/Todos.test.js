import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Todos } from '../Todos';

describe('Todos', () => {
  it('renders correctly', () => {
    const props = {
      isFetching: false,
      todos: [{ id: '123', todo: 'todo value', done: false }],
      actions: {
        addTodo: jest.fn(),
        updateTodo: jest.fn(),
        prefetchTodos: jest.fn()
      }
    };
    const tree = renderer
      .create(
        <MemoryRouter>
          <Todos {...props} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
