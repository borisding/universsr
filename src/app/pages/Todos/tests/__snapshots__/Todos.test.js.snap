// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos renders correctly 1`] = `
Array [
  <h3>
    Todos Demo
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      className="todo-input"
      name="newTodo"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Enter new todo..."
      type="text"
      value=""
    />
  </form>,
  <hr />,
  <div
    className="todos-container"
  >
    <h3>
      My Todos (0 / 1)
    </h3>
    <ul>
      <li>
        <label
          htmlFor="123"
        >
          <input
            checked={false}
            id="123"
            onChange={[Function]}
            type="checkbox"
            value="123"
          />
          <span
            className="pending"
          >
            todo value
          </span>
        </label>
      </li>
    </ul>
  </div>,
]
`;
